### DISCUSSION TITLE
Please support cross‑compiling for old ᴡɪɴᴀᴘɪ architectures.

### DISCUSSION THREAD

**Discussion by ytrezq** [open-watcom/open-watcom-v2#183] [Ideas]
Windows® and visual studio supported Alpha; PowerPC; Mips and IA64 in the past. Despite it restricts to outdated windows versions, getting those is the best way to run desktop ᴡindows® apps if you have the source code.

So I'm asking this feature request to the compiler which will  supports C11 and C++99 for eComStation modern 80286 compatible SoC _(I’m especially thinking to the ᴄᴀꜱɪᴏ one which I got last year)_.
If you do it, I’ll be able to distribute binaries for those platforms by using a single compiler with the same  version.

**Update** : I found a [Qemu way](http://gunkies.org/wiki/Installing_Windows_NT_4.0_on_Qemu%28MIPS%29) to run ᴡindows‑ɴᴛ for ᴍɪᴘꜱ on x86.

**Comment by jmalak** [comment]
What exactly you mean? It is hard because nobody is testing these platforms.
If you want to contribute by testing on these platforms you are welcome.
I don't plan to remove DOS support, but Windows 3.x support is a question.

**Comment by ytrezq** [comment]
@jmalak : That windows up to NT 3.51 supported PowerPC and up NT 4.0 Dec Alpha; MIPS _(and some others)_. Windows 2008R2 and XP were available for IA-64. Today no compiler support those platforms _(ᴍꜱ Visual droped support)_.
But if you only have those kind of machines, then using one of those Windows® version is the best way to run Software written for the ᴡɪɴᴀᴘɪ _(include win32 win16 win64 and others)_. _(please also note PowerPC PC machines are still [being produced](http://os4.hyperion-entertainment.biz/content/72/supported-hardware) by some manufacturer. I own a recently MIPS manufactured box)_. I also use Open Watcom for creating ᴡɪɴ16 graphic software for my 4MB high‑end laptop _(it use a 486 but even the linux kernel require more ram, so use windows 3.11 which have limited 32 bits support, ᴏpenwatcom is the last to support it)_.

**Reply by kristibektashi** [reply]
Technically NT 4.0 SP1 did support PowerPC, alongside x86, MIPS and DEC Alpha

**Comment by jmalak** [comment]
These platforms are probably broken, because they are untested anyway after uplift OW code to 64-bit host. It can be supported if anybody will test it. I dont have any suitable hardware and OS for testing these platforms.

**Comment by ArmstrongJ** [comment]
I was under the impression that the non-x86/x64 code generators were not in a "working" state at the moment.  The Open Watcom site specifically stated that MIPS did not work any longer.  I'd be surprised if the AXP generator worked still.  Given that Windows MIPS and PowerPC hardware was almost non-existent even in its heyday (except for, technically, the Xbox 360), supporting those platforms would be a ton of work for little to no payoff.

I'm aware that PowerPC hardware is still being produced.  However, I don't know of any currently produced PowerPC hardware that can boot Windows.  A specific firmware was necessary to start Windows.  It couldn't just start on any arbitrary PowerPC system.

Alpha Windows hardware was significantly more popular.  Again, though, like @jmalak stated, we'd need hardware or a qemu instance (already configured...) to test on.  However, since @jmalak  is focusing on x64 (and rightfully so), Alpha support might and probably should be deferred for now.

I don't think IA-64 support is a good idea at all.  The Itanium is an extremely complicated chip (VLIW), and we should definitely stay away from it.  If we want to look at supporting a completely new architecture, ARM makes far more sense than Itanium.

**Comment by davidrg** [comment]
I tried Alpha support a few years back and did get it working in v1.7.1 but getting it built involved a few manual steps to build the libraries. I attempted it in v1.9 more recently but hit some problem (probably just some minor build issue).

The AXP generator does/did actually produce programs that will run on a real Alpha making it the only generally available Alpha Windows NT compiler. I managed to build some simple Win32 apps with it and run them on one of my Alphas no problem. It can't build itself though - something missing in the code generator apparently. Might not be a huge amount of work to fix given it does work for simpler programs.

I've read the MIPS and PowerPC generators were in a much earlier stage of development when the project was canceled and so aren't finished enough to do anything useful. MIPS support would be easy enough to test in an emulator though if you can find an NT4 CD (http://virtuallyfun.superglobalmegacorp.com/2009/08/02/mips-blast-from-the-past/), PowerPC I guess might need real hardware. I've considered trying to get it going on my RS/6000 7248 but never bothered due to the PowerPC generators supposedly broken state.

**Reply by kristibektashi** [reply]
The people at Virtually Fun did get the 1.8 cross-compiler running: https://virtuallyfun.com/2010/05/27/openwatcom-for-the-dec-alpha/

**Reply by kristibektashi** [reply]
Also nowadays you can emulate NT 3.51 and 4 for PowerPC as well: https://betawiki.net/wiki/Guide:Installing_Windows_NT_on_PowerPC

**Comment by ytrezq** [comment]
@davidrg : Winworlpc distributes [ɪꜱᴏ](https://winworldpc.com/product/windows-nt-40) for all versions and platforms.
There is even a [ᴡindows 2000 ᴄᴅ](https://winworldpc.com/product/windows-nt-2000/rc-2) for ᴅᴇᴄ Alpha.
Please also send me your working Alpha build.

@ArmstrongJ : I have doubt about if supporting the new ᴀʀᴍ and ᴀʀᴍ64 is worthwhile as unless the system is hacked with a debugger, the executables _(.dll .sys nls .exe)_ need to be signed by Microsoft.

**Reply by kristibektashi** [reply]
This isn't true for ARM64, and even on ARM32 that can be easily bypassed nowadays (depending on the device, either requiring only a self-signed certificate or not requiring any signing at all)

**Comment by davidrg** [comment]
@ytrezq I can't find my AXP cross-compiler (it was quite a while ago I was playing with it) but my notes on how to build it are here: http://www.ext.zx.net.nz/software/notes/watcom-axp/. I've only tried running the binaries it generates on NT 4 SP6 - no idea what support for NT 3.x or Win2k RC2 is like.

**Comment by ytrezq** [comment]
@davidrg : [Your page](http://www.ext.zx.net.nz/software/notes/watcom-axp/) is suggesting that Microsoft Visual C++ RISC edition can work as cross‑compiler. I may be wrong but it seems RISC editions never existed as cross compilers.

**Comment by davidrg** [comment]
@ytrezq You're correct - as far as I know the RISC edition only runs on the real hardware. I've got several Alphas and at the time I was looking for a compiler to actually run on the real hardware so I wouldn't have to bother copying the generated binaries between machines whenever I wanted to test them. So Visual C++ RISC edition was the ideal choice - but its not on MSDN anymore (apparently contained the MS JVM) and too expensive second-hand. The OpenWatcom cross-compiler was my Plan B.

**Comment by ytrezq** [comment]
@davidrg : Remember there are [this](http://bingeflying.com/3.2/) and [this](https://winworldpc.com/product/visual-c/6x) for testing.

Are you sure the 386 ɴᴛ ɪᴅᴇ support alpha compile switches ? I started a project and went to the ᴄᴘᴜ flags _(`10. Memory Model and Process Switches`)_ and there are only x86 options.

**Comment by davidrg** [comment]
@ytrezq RE: compile switches - the v1.7 source release certainly includes them. It actually produced the Alpha, MIPS and PowerPC cross-compilers and included IDE support for at least the Alpha cross-compiler by default. All I had to do to make it usable was manually build all the libraries in the right order (I assume they're only excluded from the normal build to speed things up as they built fine).

As the released v1.7 binaries don't include IDE support or the cross-compilers I assume the OpenWatcom team has just been stripping them out by hand as part of the release process or there is some other build option to disable Alpha support. I wouldn't be too surprised if this has been automated in v1.8 or 1.9.

**Comment by ytrezq** [comment]
@davidrg  : Sorry I forgot to say ɪᴅᴇ not the wall compiler itself. And yes I used 1.7.0 to build 1.7.0 and still find nothing in the ɪᴅᴇ even for Alpha. Maybe are you talking about the native one under the axpnt directory ?

**Comment by davidrg** [comment]
@ytrezq I managed to find my Alpha cross compiler from back in 2012. The IDE reports itself as 1.7 and was produced with the instructions I posted earlier. The IDE has "Alpha Win32 (NT/Win95/Win32s)" as an available target (the list of supported OSes is obviously wrong). Building a quick console hello world program using that target shows wmake invokes wccaxp as the compiler and wlink including the axp libs. My x86 work PC correctly reports the resulting executable as not being a valid Win32 application so it looks like its all working. 

You should be able to download the full build here: http://ftp2.zx.net.nz/pub/misc/watcom/watcom-1.7-dec-alpha-crosscompiler.tar.xz

edit: looks like the source isn't in that archive after all. But it was just the official 1.7.1 source archive from ftp.openwatcom.org - no patches or anything.

**Comment by jmalak** [comment]
Any platform except Intel x86 was not officially released in any version of OW.
Standard installers doesn't contains anything for Alpha, PPC, etc.
But during OW full build all these parts are created that if you download (ow-snapshot.7z archive) that it should contains all what is build during OW full build (by example symbol files for debugging executables or DLLs, some support tools for Alpa, PPC ... architecture and IDE configuration file for Alpha.
But this part is untested and probably broken in OW 2.0.

**Comment by jmalak** [comment]
Anyway OW still support Win32s, NT3.x, NT4.0 targets in run-time libraries. We fixed some stuff for these versions related to getting environment.

**Comment by ytrezq** [comment]
Since we now have a working Alpha compiler which an build itself using the native architecture, maybe it would be worthwhile to upload this old binary version on Sourceforge in order to get something working.

x86 OW remained an old version for five year, so why not ?

@davidrg  : I still can’t see even with the version you builded
![ɪᴅᴇ new target](https://cloud.githubusercontent.com/assets/3824869/8798721/afed29ba-2fa4-11e5-8845-7aed2cc0a808.PNG).

**Comment by jmalak** [comment]
You are right building of ALPHA configuration file for IDE is disabled. It looks like I disabled build of most ALPHA stuff in OW 2.0.
If you are interesting in ALPHA stuff then I enable build it to be available in ow-snapshot.7z archive, but installers will not contain anything from ALPHA stuff.

**Comment by jmalak** [comment]
Good to know that OW 1.7.1 alpha compiler is working.
We can use it as reference for generated code.

**Comment by ytrezq** [comment]
@jmalak : The picture come from the link above : http://ftp2.zx.net.nz/pub/misc/watcom/watcom-1.7-dec-alpha-crosscompiler.tar.xz _(1.7.1)_ I only see command line compilers.

“`We can use it as reference for generated code.`” and probably upload it to sourceforge _(thought we can add the native alpha compiler in the same package)_ ?

**Comment by jmalak** [comment]
Our code base must be consistent that we can not do this way. We should fix Alpha code generator/compiler if it is broken. For checking generated code we can compare results with OW 1.7.1.
If you are able to check OW 2.0 on Alpha system it is best test.

**Comment by davidrg** [comment]
@ytrezq I suspect you're not seeing the Alpha target because you've got another version of OpenWatcom in your path. On my work machine the only copy I have is the binaries from that tar.xz file and this is the targets dialog I get:
![171-alpha-screenshot](https://cloud.githubusercontent.com/assets/1887527/8811842/b907c962-304c-11e5-836b-7456d8db750b.png)

As for a native Alpha compiler - that isn't something that currently exists. The x86 compiler can build an x86 version the Alpha compiler but the Alpha compiler can not build the Alpha compiler due to some missing functionality (Michal Necasek suggested it was missing at least support for unsigned long long <-> double conversion in the code generator). So to get a native Alpha compiler you'd either need to build it with Visual C++ 6.0 RISC edition or fix whatever is preventing it from building itself.

IIRC the Alpha libraries don't build in OpenWatcom 1.9 and I assume 2.0 so of course that would require fixing first in order to get a newer cross-compiler

**Comment by ytrezq** [comment]
At least I can confirm I managed to build wpp386.exe and wppaxp.exe cl.exe cl386.exe claxp.exe clppc.exe link.exe link386.exe linkaxp.exe linkppc.exe lib.exe libaxp.exe libppc.exe wasm.exe wasppc.exe rc.exe nmake.exe cvtres.exe and some others for the alpha instruction set.

Though, I still can’t build my dll for the native subsystem.

**Comment by jmalak** [comment]
cl...exe, link...exe and lib...exe are only stubs to OW tools.
But wlib.exe, wlink.exe, wccaxp.exe and wppaxp.exe are important.
Are you able to check compile and link some small program by OW2.0 tools and run it on Alpha?

**Comment by ytrezq** [comment]
Like my previous comment is saying I can create all the executables of the`axpnt` directory. I don’t have alpha hardware but getting it running on qemu seems only to be the matter of a [trivial fix](https://bugs.launchpad.net/qemu/+bug/1473451).
ʙecause of course @davidrg already build the cross compiler version of wppaxp.exe.

Since alpha ᴡindows® is able to run x86 ᴡɪɴ32 binaries without difference from native ones, I think wpp386.exe is more important than wppaxp.exe as the purpose would be to use servers for building projects using less resources.
The real meaning of creating alpha native code for windows would be to cover code which can’t use the x86 instruction set : executables for the native subsystem.

**Comment by davidrg** [comment]
@ytrezq Alpha Windows can only run 16bit x86 binaries in its built-in emulator. Trying to run 32bit x86 binaries will get you the "isn't a valid win32 application" error message unless you have FX!32 (a 3rd party translator/emulator from DEC) installed.

As for native binaries, last time I tried to build them the compiler died with internal error 92 in /bld/cc/c/cfold.c. The problematic lines were apparently the ones #ifdefd with __GNUC__ such as https://github.com/open-watcom/open-watcom-v2/blob/1b9e0ce0d53e289b194120e3e1cba76c6cab867e/bld/cc/c/cfold.c#L572. If you're able to build newer version of this file I guess the problematic lines must have been removed or otherwise changed but I expect the features required to build code like this are still missing in the Alpha code generator.

**Comment by joncampbell123** [comment]
I would like to add that I do Windows 3.x development as part of my DOSLIB project. I use DOSBox and QEMU to run Windows 3.0 and Windows 3.1 to test the code. I can help with making sure Windows 3.x support still works because OW is (as far as I know) the only open source compiler that supports Windows 3.x as a target so well.

**Comment by jmalak** [comment]
It will be great if you are able to do time to time checking of Windows 3.x support and report any problem which you have that anybody can fix it.

**Comment by joncampbell123** [comment]
Good.

**Comment by ytrezq** [comment]
@joncampbell123 : I opened this issue only about supporting other architectures on  other ᴡindows®.

**Comment by joncampbell123** [comment]
Yes. I guess I misread some of the postings here as suggesting that Windows 3.x support would be dropped from Open Watcom.

**Comment by ytrezq** [comment]
Interesting docs
ftp://ftp.openwatcom.org/pub/devel/docs/powerpc%20compiler%20writers%20guide.pdf
ftp://ftp.openwatcom.org/pub/devel/docs/alpha%2021264%20compiler%20writers%20guide.pdf
ftp://ftp.openwatcom.org/pub/devel/docs/alpha%20architecture%20handbook.pdf
ftp://ftp.openwatcom.org/pub/devel/docs/alpha%20architecture%20handbook.pdf
ftp://ftp.openwatcom.org/pub/devel/docs/alpha%20nt%20porting%20from%20x86.pdf
ftp://ftp.openwatcom.org/pub/devel/docs/alpha_arch_ref.pdf
ftp://ftp.openwatcom.org/pub/devel/docs/ia64abi.pdf

ᴡindows® supported 2 MIPS ISA as separated architecture _(eg you can’t run MIPS ISA III binaries on ISA IV)_. ftp://ftp.openwatcom.org/pub/devel/docs/r4000%20microprocessor%20users%20manual.pdf ftp://ftp.openwatcom.org/pub/devel/docs/r10000%20microprocessor%20users%20manual.pdf

@davidrg : shall I understand there is no 64 bits support for ᴅᴇᴄ Alpha ?

**Comment by davidrg** [comment]
@ytrezq All public releases of Windows for the Alpha (including the Windows 2000 RCs) are 32bit only. 64bit Windows 2000 did exist for the Alpha at one point early on in the 64bit Itanium ports development as real Itanium hardware didn't exist at the time but as far as I know none of these builds ever escaped Microsoft.

On the MIPS side, as far as I know only R4x00 series (MIPS III) processors were ever supported by the retail releases. MIPS support was discontinued entirely after NT 4.0 service pack 1 (PowerPC survived just long enough to see service pack 2).

**Reply by kristibektashi** [reply]
Well nowadays it does exist: https://betawiki.net/wiki/Windows_XP_build_2210 and https://virtuallyfun.com/2023/05/15/windows-2000-64-bit-for-alpha-axp/

And there's even a compiler and linker included in the Windows 2000 RC2 SDK: https://virtuallyfun.com/2023/05/05/hiding-in-plain-sight-the-64bit-dec-alpha-c-compiler/ and https://archive.org/details/ms-win2k-platsdk-rc2. 

This is a cross-compiler that only runs on 32-bit Alpha (if you install on x86 all you get is a broken Itanium compiler) and can compile stuff for AXP64, all the libraries and stuff are also there I think. And you can run any compiled executables on the Windows build I mentioned above, assuming you have a real DEC Alpha machine

**Comment by ytrezq** [comment]
@davidrg : I did really setup a system that can run and build ᴍɪᴘꜱ ɪꜱᴀ ɪᴠ executables _(refusing to run ɪꜱᴀ ɪɪɪ)_. The visual c++ linker can’t generate export _(.exp)_ files though, requiring manual writing. But once done, the dlls _(including kernel ones)_ I wrote work.

Are you sure for 64 bits ᴅᴇᴄ Alpha ? 64 axp bits executables are defined in the [PE executable format specification](http://download.microsoft.com/download/e/b/a/eba1050f-a31d-436b-9281-92cdfeae4b45/pecoff.doc) from MS.

**Comment by davidrg** [comment]
@ytrezq I guess Microsoft may have planned MIPS IV support at some point which would explain the limited compiler support (the first chip appeared in 1994) but my copies of NT all say they require a MIPS R4x00. Visual C++ support for MIPS IV may also be for building Windows CE applications rather than Windows NT ones. Given special firmware is required to run windows (you wont, for example, get it to run on an SGI) I don't think any actual windows-capable MIPS IV machine was ever released.

Windows in general was 32bit regardless of CPU architecture until Windows Datacenter Server Limited Edition was released in 2002. This was the first 64bit version which ran on the Itanium and was based on the at the time unreleased Windows Sever 2003 codebase.

**Comment by ytrezq** [comment]
@davidrg : Acutually putting a ᴍɪᴘꜱ ɪꜱᴀ ɪᴠ ᴄᴘᴜ on ᴍɪᴘꜱ magnum works. Also Windows ᴄᴇ use ɪꜱᴀ ɪɪ not ɪɪɪ.

but what about AXP64 ?

**Comment by davidrg** [comment]
If windows works on a MIPS IV I guess its just because the chip is sufficiently backwards compatible - it doesn't appear to have ever been a supported configuration.

As for AXP64, Alpha was a 64bit architecture from the start - there was never a 32bit variant of it. So when you're running Windows NT on an Alpha you're running a 32bit operating system on a 64bit CPU.

**Comment by ytrezq** [comment]
@davidrg : No, I also saw some peoples producing 64 bits executables as well as 32 bits ones for the ᴅᴇᴄ Alpha architecture on windows. A good example is the [dependency walker](http://www.dependencywalker.com/) included in many Microsoft® products.

**Comment by davidrg** [comment]
@ytrezq All executables on an Alpha are 64bit because the Alpha has no 32bit mode - its a 64bit processor. But the Win32 API only uses 32bit pointers into a 32bit virtual address space making it a 32bit operating system. While the hardware can see more than 4GB of RAM windows will never make use of it or allocate it to user-mode applications.

**Comment by ytrezq** [comment]
@davidrg : But as you can see it on [that page](http://www.dependencywalker.com/) there are the two executable version and the machine type inside the PE header is different _(Alpha and AXP64)_. This is made similar with ɪᴀ‑32 and ᴀᴍᴅ64.

**Comment by davidrg** [comment]
@ytrezq I guess its possible the AXP64 one might have been for the never-released internal 64bit Alpha version of Windows Server 2000. This is discussed a little here: https://technet.microsoft.com/en-us/magazine/2008.08.windowsconfidential.aspx.

There was actually a special way for Windows applications to allocate memory outside of Windows' 32-bit virtual address space but about the only thing it was really useful for was disk I/O. The Win32 API still only accepted 32bit pointers making the 64bit pointers returned by VirtualAllocVlm fairly useless otherwise. VirtualAllocVlm and Very Large Memory support is discussed here: http://blogs.msdn.com/b/oldnewthing/archive/2007/08/01/4155462.aspx

My guess is the AXP64 binary is either mislabeled or its for an operating system that was never publicly released.

**Comment by ytrezq** [comment]
@davidrg : [The file](http://www.dependencywalker.com/depends20_axp64.zip) contains 64 bits instructions. Also several projects are in that case.

**Comment by ArmstrongJ** [comment]
Windows NT on Alpha was, without question, a 32-bit operating system.  64-bit NT was being planned with the release of Windows 2000 on Alpha, but it was canceled.  [Wikipedia states pretty clearly](https://en.wikipedia.org/wiki/Windows_NT#64-bit_platforms) that all versions for the Alpha are 32-bit.  Furthermore, a bit of searching will reveal plenty of articles from reputable publications also explaining that Windows NT 5.0, eventually called Win2k, would be the first to support 64-bit in any form.  Since it was never released for Alpha, it doesn't "exist" in a supportable form.  It did exist inside Microsoft, but that's it.

The architecture name of AXP64 is commonly used, so I would imagine that is why you're seeing files for 32-bit Windows NT labeled as such in a misleading manner.

Again, all released versions of Windows NT on Alpha are 32-bit.

**Comment by ytrezq** [comment]
Since peoples can’t click on a link.
![](https://cloud.githubusercontent.com/assets/3824869/9122275/2d2b8d78-3c88-11e5-8a45-ff1f127d78f5.PNG)Executables details _(look at the`Preferred Base`column)_ :![](https://cloud.githubusercontent.com/assets/3824869/9122914/71131430-3c8c-11e5-8e5d-d090847c85b6.PNG)![](https://cloud.githubusercontent.com/assets/3824869/9122913/710eab02-3c8c-11e5-80b5-84beed4aa8e0.PNG)
@ArmstrongJ :  Other projects are in that case _(version for 32 and 64 bits ᴅᴇᴄ Alpha)_.

**Comment by ArmstrongJ** [comment]
@ytrezq I see that someone compiled a version for 64-bit Windows for Alpha, or so it appears.  However, that doesn't change the fact that a 64-bit version of Windows NT was _never released_ for the Alpha architecture.  NT on Alpha died with NT4, and that operating system was 32-bit.  There were _development_ versions of 64-bit NT for Alpha, but that's it.

With the manpower in this project, we'd be lucky to get the Alpha compiler and Windows API up and running (again, I'm sure it worked in the past) for 32-bit, if at all.  Asking for support on other NT architectures, especially never-released 64-bit Win2k for Alpha, is really a stretch.

If you're really interested in this, why don't you contribute to getting the Alpha 32-bit builds and the related Windows API functioning?  I'd have no problem with seeing Open Watcom work there again (if it doesn't already, I haven't tried it).

**Comment by ytrezq** [comment]
@ArmstrongJ : I can just do to testing. As a student, I don’t have the required knowledge for working with assembly on modern ᴄᴘᴜ architectures yet.

**Comment by pchapin** [comment]
@ArmstrongJ Open Watcom definitely used to work on Windows NT Alpha, and it wasn't _too_ long ago. I know Michal had Open Watcom self compiling on NT Alpha at one point. There were a few issues, as I recall, but he was able to compile non-trivial programs in that environment.

**Comment by galazwoj** [comment]
Hi, I managed to compile wccaxp and run 'pmake axp' over clib and mathlib . Wccaxp was used as a crosscompiler to build alpha libraries, the host OS is NT386. Now I'm progressing to build a native compiler from bld\cc\axp\ntaxp.dll.  Once done I will try to check it out in a virtual env (qemu?) as I haven't used my real Alpha hardware for years. Let me know how could I submit my changes. I don't know git much :(

**Comment by jmalak** [comment]
standard POSIX diff output is acceptable if you have trouble with git.
But you can use some git frontend as Syntevo SmartGit etc. to works with git.

**Comment by galazwoj** [comment]
Here are the patches made as diff -c. 
- compiler.patch fixes wccaxp errors when compiling code for NTAXP target 
- linker.patch fixes wlink errors when linking NTAXP programs
- clib-alpha.patch fixes clib errors when compiling with wccaxp
- math-alpha.patch fixes mathlib errors when compiling with wccaxp

[math-alpha.patch.txt](https://github.com/open-watcom/open-watcom-v2/files/900653/math-alpha.patch.txt)
[clib-alpha.patch.txt](https://github.com/open-watcom/open-watcom-v2/files/900651/clib-alpha.patch.txt)
[compiler.patch.txt](https://github.com/open-watcom/open-watcom-v2/files/900652/compiler.patch.txt)
[linker.patch.txt](https://github.com/open-watcom/open-watcom-v2/files/900654/linker.patch.txt)

I haven't tested native AXP programs on Windows NT Alpha yet.

Comments welcome

**Comment by jmalak** [comment]
Thanks for your contribution.
Give me some time to apply them.
I must convert your patches to unix format to be able apply them and review your changes.
I will return as soon as I will apply all locally or submit something into repository directly.

**Comment by jmalak** [comment]
There are changes in kernel32 entry points, added entry points for 64-bit Windows
RtlLookupFunctionEntry
RtlUnwindRfp

It is unacceptable, because this import library is for 32-bit Windows only.
I didn't applied this change.

I reworked a few of your changes in math run-time library related to error handling and submit all math library changes.

I have a question related to changes in C compiler.
You use macro `_INT64_BROKEN_`, what is broken on 64-bit integers?

I will review rest of your changes ASAP.

**Comment by galazwoj** [comment]
Thanks for looking into the patches

The changes in kernel32 entry points are for two functions present in Windows NT Alpha.  I don't think there can be other header file they can be placed in. Without these entry points clib cannot be compiled. 

The macro _INT64_BROKEN_, is a crude attempt to work around the lack of casting from long long to double in the Alpha compiler. The other way is to add the missing functionality into code generator which I do know how to do now.  When  _INT64_BROKEN_ is not defined any attempt to cast from long long to double variable results in wccaxp internal error 93 as far as I can recall.

**Comment by jmalak** [comment]
You are right. I checked old documentation for Windows NT AXP Edition and these two symbols are specific for Alpha. I will add them to kernel32 Import library.
I will try to look to Alpha Code generator for this issue and fix it if I was able to identify source of problem.

**Comment by jmalak** [comment]
I added new Windows NT functions specific for Alpha into Alpha kernel32 import library and added prototypes for all Alpha specific function into winnt.h header file.
I did appropriate changes to use new prototypes in C and C++ run-time libraries.

**Comment by galazwoj** [comment]
Today I went up to the point where 
- pmake os_nt cpu_386 .and target_ppc .not javavm .not 
- pmake os_nt cpu_axp .and target_ppc .not javavm .not trap .not

started in bld directory run without  errors
Unfortumately PPC target is incomplete, javavm requires some sort of Java SDK and trap on AXP requires some time to understand. 
I will incorporate your changes and then perform more testing to make sure all this works and do not breaks other stuff and then can send some diff -u3 patches. I use diff from mingw/msys

**Comment by jmalak** [comment]
What is interesting that if I build Alpha compiler on 32-bit Windows then it works without problem with 64-bit integers without any correction.
It looks like portability issue for 32/64-bit hosts.

**Comment by galazwoj** [comment]
When I try to compile the original bld\cc\c\cfold.c using wccaxp built with the same original cfold.c on Windows XP 32 then the compiler crashes.  Other people reported the same problem in the past,, for example here are some comments.
https://virtuallyfun.superglobalmegacorp.com/2010/05/27/openwatcom-for-the-dec-alpha/

The directory is bld\cc\axp\nt386.dll
After patching the file the compiler generates no errors.

**Comment by jmalak** [comment]
I did a little investigation and a problem is deeper then it looks like.
Code generator doesn't support double <-> 64-bit unsigned integer conversion.
Original Watcom code supported only 32-bit integer therefore no problem in the past.
We need some general solution, because Alpha processor hasn't CPU instruction for this conversion.
Your solution fix only C compiler and partialy.
Simple solution could be to use Alpha double <-> 64-bit signed integer conversion and support conversion for positive numbers only.
Next  possibility is to use C run-time library support function for this conversion and change code generator to generate appropriate call.

**Comment by jmalak** [comment]
I did last commit 3063a01f37a5891e5011473a0362af38d0e55ba7 with temporary fix to be able compile Alpha Compiler for Alpha platform.
But it still requires fix code generator for missing unsigned integer <-> float conversions

**Comment by galazwoj** [comment]
I wanted to check what code was generated by Microsoft C++ for AXP. Unfortunately
it is not yet possible to run Windows AXP in qemu emulator or any other DEC Alpha emulator 
so I'm going to run my alpha workstation but not sooner than in July in order to find the results.
Out of curiosity I tried gcc in emulated Debian Alpha but the code is overly complex.

Interestingly Watcom C++ compiler can be compiled and it works without any specific patches.
Both wccaxp and wppaxp generate exactly the same code for small test cases tried.
To convert form long long to double a pair of instructions is used (registers may vary)
`	ldt	$f18,($sp)	;load quadword integer into floating point register`
`	cvtqt	$f18,$f18	;convert quadword to T_floating (that is to IEEE double) `

**Comment by jmalak** [comment]
Only C compiler uses unsigned integer <-> float conversion. After fixing it uses only signed integer <-> float conversion.
C++ compiler can be compiled without problem because it doesn't use any unsigned long long <-> float conversion.

**Comment by galazwoj** [comment]
You're right, Just a quick test trying to compile a simple case
void print1u(unsigned long long x)
`{`
`        double xx = x;`
`        printf("value is  %lld, %lf", x, xx);`
`}`
results in Error! E1119: Internal compiler error 92
both for wppaxp and wccaxp
I will dig more into the possible solution.

**Comment by galazwoj** [comment]
gcc does the conversion in a way than can be recoded in C as follows
`double U8FD(unsigned long long i)`
`{`
`	if (i <= LONG_LONG_MAX) {`
`		return (double)((signed long long)i);`
`	} else {`
`		signed long long ll = (signed long long)(i >> 1)`;
`		signed long long  r = (signed long long)(i & 1)`;
`		double d = ll;`
`		d = d + d + r;`
`		return d;`
`	}`
`}`
Code generated by Watcom seems to be good enough that no assembly coding seems necessary

Example usage
`void display(unsigned long long x)`
`{`
`	double xx = U8FD(x);`
`	printf("value is %llu, %f\n", x, xx);`
`}`

Perhaps the easiest way is to use this as C run-time library support function and change code generator to generate appropriate call to make the conversion..

**Comment by jmalak** [comment]
Thanks for analysis.
I will look if it is possible to generate this code in-line or if it needs to use C run-time library support function.

**Comment by jmalak** [comment]
One question.
Do you know how gcc implement double to unsigned long long conversion on Alpha?

**Comment by galazwoj** [comment]
I can check this and will post the results.

**Comment by galazwoj** [comment]
gcc implements double to unsigned long long conversion exactly in the same way as conversion from double to signed long long,, the assembly code generated is 100 % the same..
I see no reason why OW could not do this as well.
I did little tests with wccaxp and enclose the results for those who do not have wccaxp working.

[all2x.axp.txt](https://github.com/open-watcom/open-watcom-v2/files/1033504/all2x.axp.txt)
[all2x.c.txt](https://github.com/open-watcom/open-watcom-v2/files/1033505/all2x.c.txt)

**Comment by galazwoj** [comment]
There is a subtle bug somewhere in cg which is axp related. It manifests in infinite loop when optimization is tuned on but does not show up when -oaxp is not specified. I am not sure where to ask how to track down the error. So far I managed to create a sample source file that  activates the bug.. This bug is related to register allocation,, there seems to be not enough registers and the optimizer creates larger and larger number of temporary variables in each optimization step
